numpy/ma/core.py

```diff
@@ -4286,22 +4297,45 @@ def count(self, axis=None):
         array([3, 0])
 
         """
+        kwargs = {} if keepdims is np._NoValue else {'keepdims': keepdims}
+
         m = self._mask
-        s = self.shape
+        # special case for matrices (we assume no other subclasses modify
+        # their dimensions)
+        if isinstance(self.data, np.matrix):
+            if m is nomask:
+                m = np.zeros(self.shape, dtype=np.bool_)
+            m = m.view(type(self.data))
+
         if m is nomask:
-            if axis is None:
+            # compare to _count_reduce_items in _methods.py
+
+            if self.shape is ():
+                return 1
+            elif axis is None:
                 return self.size
+
+            axes = axis if isinstance(axis, tuple) else (axis,)
+            items = 1
+            for ax in axes:
+                items *= self.shape[ax]
+
+            if kwargs.get('keepdims', False):
+                out_dims = list(self.shape)
+                for a in axes:
+                    out_dims[a] = 1
             else:
-                n = s[axis]
-                t = list(s)
-                del t[axis]
-                return np.full(t, n, dtype=np.intp)
-        n1 = np.size(m, axis)
-        n2 = np.sum(m, axis=axis, dtype=np.intp)
-        if axis is None:
-            return (n1 - n2)
-        else:
-            return narray(n1 - n2)
+                out_dims = [d for n,d in enumerate(self.shape) if n not in axes]
+            # make sure to return a 0-d array if axis is supplied
+            ret = np.empty(out_dims, dtype=np.intp)
```

Comment:
Again sorry to say this only now, but why not use `np.full` as in the original? (Really a detail, so feel free to ignore).